File: crypto_tax/src/exporter.py
"""
Exporter: converts a TaxReport into TurboTax-compatible CSV files.

TurboTax accepts cryptocurrency transactions in a CSV with these columns
(Form 8949 / Schedule D layout):

  Currency Name     – description of the asset (e.g. "1.5 BTC")
  Purchase Date     – date acquired  (MM/DD/YYYY)
  Cost Basis (USD)  – your cost in USD
  Date Sold         – date disposed  (MM/DD/YYYY)
  Proceeds (USD)    – amount received in USD

Income (staking rewards, interest, cashback, etc.) is reported on Schedule 1
(Other Income) and exported to a separate CSV with columns:

  Currency Name     – asset symbol
  Date Received     – date of receipt (MM/DD/YYYY)
  Amount Received   – quantity of crypto
  Fair Market Value (USD) – value in USD on receipt date
  Income Type       – description / transaction kind
"""

import csv
import io
from typing import Tuple

from .calculator import TaxReport

DATE_FMT = "%m/%d/%Y"


def export_turbotax_csv(report: TaxReport) -> Tuple[str, str]:
    """
    Generate two CSV strings suitable for TurboTax upload.

    Returns
    -------
    (capital_gains_csv, income_csv) : Tuple[str, str]
        capital_gains_csv – Form 8949 / Schedule D disposals
        income_csv        – Schedule 1 ordinary-income receipts
    """
    # ── Capital-gains CSV (Form 8949) ─────────────────────────────────────────
    cg_buf = io.StringIO()
    cg_writer = csv.writer(cg_buf)
    cg_writer.writerow(
        [
            "Currency Name",
            "Purchase Date",
            "Cost Basis (USD)",
            "Date Sold",
            "Proceeds (USD)",
            "Gain or Loss (USD)",
            "Term",
        ]
    )

    for evt in report.tax_events:
        asset_label = f"{evt.quantity:.8g} {evt.currency}"
        cg_writer.writerow(
            [
                asset_label,
                evt.date_acquired.strftime(DATE_FMT),
                f"{evt.cost_basis_usd:.2f}",
                evt.date_sold.strftime(DATE_FMT),
                f"{evt.proceeds_usd:.2f}",
                f"{evt.gain_loss_usd:.2f}",
                "Long-term" if evt.is_long_term else "Short-term",
            ]
        )

    # ── Income CSV (Schedule 1) ───────────────────────────────────────────────
    inc_buf = io.StringIO()
    inc_writer = csv.writer(inc_buf)
    inc_writer.writerow(
        [
            "Currency Name",
            "Date Received",
            "Amount Received",
            "Fair Market Value (USD)",
            "Income Type",
        ]
    )

    for evt in report.income_events:
        inc_writer.writerow(
            [
                evt.currency,
                evt.date_received.strftime(DATE_FMT),
                f"{evt.quantity:.8g}",
                f"{evt.fair_market_value_usd:.2f}",
                evt.description,
            ]
        )

    return cg_buf.getvalue(), inc_buf.getvalue()


def export_summary_txt(report: TaxReport, method: str) -> str:
    """Return a human-readable plain-text summary of the tax report."""
    lines = [
        "=" * 60,
        "  CRYPTO TAX SUMMARY",
        f"  Method: {method}",
        "=" * 60,
        "",
        f"  Short-term capital gains/losses : ${report.total_short_term_gain_loss:>12.2f}",
        f"  Long-term  capital gains/losses : ${report.total_long_term_gain_loss:>12.2f}",
        f"  Total capital gains/losses      : ${report.total_gain_loss:>12.2f}",
        f"  Ordinary income (crypto)        : ${report.total_ordinary_income:>12.2f}",
        "",
        f"  Total taxable events  : {len(report.tax_events)}",
        f"  Total income events   : {len(report.income_events)}",
        "=" * 60,
    ]
    return "\n".join(lines)

File: products/crypto_tax_serverless/src/exporter.js
/**
 * Exporter: converts a TaxReport into TurboTax-compatible CSV strings.
 *
 * TurboTax accepts cryptocurrency transactions in a CSV with these columns
 * (Form 8949 / Schedule D layout):
 *
 *   Currency Name     – description of the asset (e.g. "1.5 BTC")
 *   Purchase Date     – date acquired  (MM/DD/YYYY)
 *   Cost Basis (USD)  – your cost in USD
 *   Date Sold         – date disposed  (MM/DD/YYYY)
 *   Proceeds (USD)    – amount received in USD
 *   Gain or Loss (USD)– gain or loss
 *   Term              – "Long-term" or "Short-term"
 *
 * Income (staking rewards, interest, cashback, etc.) is reported on Schedule 1
 * (Other Income) and exported to a separate CSV with columns:
 *
 *   Currency Name          – asset symbol
 *   Date Received          – date of receipt (MM/DD/YYYY)
 *   Amount Received        – quantity of crypto
 *   Fair Market Value (USD)– value in USD on receipt date
 *   Income Type            – description / transaction kind
 */

const DATE_FMT_OPTS = { month: "2-digit", day: "2-digit", year: "numeric" };

/**
 * Format a Date as MM/DD/YYYY using UTC values.
 * @param {Date} date
 * @returns {string}
 */
function formatDate(date) {
  const mm = String(date.getUTCMonth() + 1).padStart(2, "0");
  const dd = String(date.getUTCDate()).padStart(2, "0");
  const yyyy = date.getUTCFullYear();
  return `${mm}/${dd}/${yyyy}`;
}

/**
 * Escape a value for CSV (wrap in quotes if it contains commas, quotes, or newlines).
 * @param {string|number} val
 * @returns {string}
 */
function csvEscape(val) {
  const s = String(val);
  if (s.includes(",") || s.includes('"') || s.includes("\n")) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}

/**
 * Build a CSV row string.
 * @param {Array} fields
 * @returns {string}
 */
function csvRow(fields) {
  return fields.map(csvEscape).join(",");
}

/**
 * Generate TurboTax-compatible CSV strings from a TaxReport.
 *
 * @param {import('./calculator').TaxReport} report
 * @returns {{ capitalGainsCsv: string, incomeCsv: string }}
 */
function exportTurboTaxCsv(report) {
  // ── Capital-gains CSV (Form 8949 / Schedule D) ────────────────────────────
  const cgRows = [
    csvRow([
      "Currency Name",
      "Purchase Date",
      "Cost Basis (USD)",
      "Date Sold",
      "Proceeds (USD)",
      "Gain or Loss (USD)",
      "Term",
    ]),
  ];

  for (const evt of report.taxEvents) {
    const assetLabel = `${fmtQty(evt.quantity)} ${evt.currency}`;
    cgRows.push(
      csvRow([
        assetLabel,
        formatDate(evt.dateAcquired),
        evt.costBasisUsd.toFixed(2),
        formatDate(evt.dateSold),
        evt.proceedsUsd.toFixed(2),
        evt.gainLossUsd.toFixed(2),
        evt.isLongTerm ? "Long-term" : "Short-term",
      ])
    );
  }

  const capitalGainsCsv = cgRows.join("\n") + "\n";

  // ── Income CSV (Schedule 1) ───────────────────────────────────────────────
  const incRows = [
    csvRow([
      "Currency Name",
      "Date Received",
      "Amount Received",
      "Fair Market Value (USD)",
      "Income Type",
    ]),
  ];

  for (const evt of report.incomeEvents) {
    incRows.push(
      csvRow([
        evt.currency,
        formatDate(evt.dateReceived),
        fmtQty(evt.quantity),
        evt.fairMarketValueUsd.toFixed(2),
        evt.description,
      ])
    );
  }

  const incomeCsv = incRows.join("\n") + "\n";

  return { capitalGainsCsv, incomeCsv };
}

/**
 * Generate a human-readable plain-text summary.
 *
 * @param {import('./calculator').TaxReport} report
 * @param {string} method
 * @returns {string}
 */
function exportSummaryTxt(report, method) {
  const lines = [
    "============================================================",
    "  CRYPTO TAX SUMMARY",
    `  Method: ${method}`,
    "============================================================",
    "",
    `  Short-term capital gains/losses : ${fmtUsd(report.totalShortTermGainLoss)}`,
    `  Long-term  capital gains/losses : ${fmtUsd(report.totalLongTermGainLoss)}`,
    `  Total capital gains/losses      : ${fmtUsd(report.totalGainLoss)}`,
    `  Ordinary income (crypto)        : ${fmtUsd(report.totalOrdinaryIncome)}`,
    "",
    `  Total taxable events  : ${report.taxEvents.length}`,
    `  Total income events   : ${report.incomeEvents.length}`,
    "============================================================",
  ];
  return lines.join("\n");
}

function fmtQty(qty) {
  // Use up to 8 significant digits, no trailing zeros
  return parseFloat(qty.toPrecision(8)).toString();
}

function fmtUsd(amount) {
  const sign = amount < 0 ? "-" : "";
  return `${sign}$${Math.abs(amount).toFixed(2)}`;
}

// Export for Node.js (tests) and browser (global)
if (typeof module !== "undefined" && module.exports) {
  module.exports = { exportTurboTaxCsv, exportSummaryTxt, formatDate, csvEscape, csvRow };
}
